### PR TITLE
[FEATURE: UIPT-19] - Slideshow

### PR DESCRIPTION
**Pull request by andriiveremchuk**

### CODE REVIEW

File: src/app/app.component.spec.ts [hunk 1]
```diff
@@ -1,12 +1,62 @@
 import { TestBed, async } from '@angular/core/testing';
+import { NgModule } from '@angular/core';
+import { BrowserModule } from '@angular/platform-browser';
+import { FormsModule } from '@angular/forms';
+import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
+import { HttpService } from './http.service';
+import { HttpClientModule } from '@angular/common/http';
+import { FontAwesomeModule } from '@fortawesome/angular-fontawesome';
+
+
```
Comment (AlexLarusso) [inline src/app/app.component.spec.ts]: remove multiple empty lines

File: src/app/slideshow/slideshow.component.spec.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    fakePaginationClick = new MouseEvent('click', { bubbles: true });
+    component.ngAfterViewInit();
+  });
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
+  });
+
+  it('click on the next arrow button should change slide to the left', () => {
+    const fakeNextClick = new MouseEvent('click', { bubbles: true });
```
Comment (AlexLarusso) [inline src/app/slideshow/slideshow.component.spec.ts]: empty line after const declarations

File: src/app/slideshow/slideshow.component.ts [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+} from 'rxjs/operators';
+import { slideshowAnimation } from './slideshow.animations';
+
+import '../../assets/img/beauty-casual-curly.jpg';
+import '../../assets/img/blazers-daytime-dress.jpg';
+import '../../assets/img/bored-boredom-casual.jpg';
+import '../../assets/img/city-daylight-diversity.jpg';
+
+const images: string[] = [
+  '../../assets/img/beauty-casual-curly.jpg',
```
Comment (AlexLarusso) [inline src/app/slideshow/slideshow.component.ts]: DRY:
Move urls like `'../../assets/img/beauty-casual-curly.jpg'` to const

File: src/app/slideshow/slideshow.component.ts [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+})
+export class SlideshowComponent implements OnInit, AfterViewInit {
+  @ViewChild('previous', { static: true }) previousEl: ElementRef;
+  @ViewChild('next', { static: true }) nextEl: ElementRef;
+  @ViewChild('slider', { static: true }) sliderEl: ElementRef;
+  @ViewChild('bullet', { static: true }) bulletEl: ElementRef;
+  public images: Array<any> = images;
+  public currentIndex = 0;
+  public currentDirection = 'left';
+  private isOnSlider = new EventEmitter<boolean>();
```
Comment (AlexLarusso) [inline src/app/slideshow/slideshow.component.ts]: 1) Don't mix private and public methods
2) ViewChilds need types too
3) Separate private, public and ViewChilds using empty lines

File: src/app/slideshow/slideshow.component.ts [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+  public timerSub: any;
```
Comment (AlexLarusso) [inline src/app/slideshow/slideshow.component.ts]: it is of type Subscription from rxjs

File: src/app/slideshow/slideshow.component.ts [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+  public clicked = false;
+
+  constructor() {}
```
Comment (AlexLarusso) [inline src/app/slideshow/slideshow.component.ts]: Don't use empty construtors

File: src/app/slideshow/slideshow.component.ts [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+          takeWhile(() => isOnSlider && !this.clicked),
+          map(() => ({ shift: 1, direction: 'left' }))
+        )
+      )
+    );
+
+    fromEvent(this.sliderEl.nativeElement, 'mouseover').subscribe(
+      () => {
+        this.isOnSlider.emit(false);
+      }
```
Comment (AlexLarusso) [inline src/app/slideshow/slideshow.component.ts]: use one line arrow function. Check everywhere

File: src/app/slideshow/slideshow.component.ts [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+        startWith({ index: 0 } as any),
+        scan((acc, curr) => {
+          const projectedIndex = acc.index + curr.shift;
+
+          const adjustedIndex =
+            projectedIndex < 0
+              ? this.images.length - 1
+              : projectedIndex >= this.images.length
+              ? 0
+              : projectedIndex;
```
Comment (AlexLarusso) [inline src/app/slideshow/slideshow.component.ts]: Rewrite not using nesting ternary operator

File: src/app/slideshow/slideshow.component.ts [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+      )
+      .subscribe(event => {
+        this.currentIndex = event.index;
+        this.currentDirection = event.direction;
+      });
+  }
+
+  public ngAfterViewInit(): void {
+    const event = new MouseEvent('mouseout', { bubbles: true });
+    this.sliderEl.nativeElement.dispatchEvent(event);
```
Comment (AlexLarusso) [inline src/app/slideshow/slideshow.component.ts]: empty line after const. Check everywhere

File: src/app/slideshow/slideshow.component.ts [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+  }
+
+  public toggleSlide(num: number): void {
```
Comment (AlexLarusso) [inline src/app/slideshow/slideshow.component.ts]: `num: number` -> name it more properly

File: src/app/slideshow/slideshow.component.ts (hunk 4 above)
Comment (andriiveremchuk) [reply src/app/slideshow/slideshow.component.ts]: Isn't an ElementRef type of a ViewChild?

File: src/app/slideshow/slideshow.component.ts (hunk 5 above)
Comment (andriiveremchuk) [reply src/app/slideshow/slideshow.component.ts]: it is of type ObservableInput<unknown> :)

**Pull request merged**